src/mono/mono/mini/interp/transform.c

```diff
@@ -8364,49 +8373,141 @@ interp_mark_reachable_bblocks (TransformData *td)
+		*cond_ins = second;
+		return TRUE;
+	}
+	// pattern `ldc; binop conditional branch`
+	if (first && second 
+		&& MINT_IS_LDC(first->opcode)
+		&& MINT_IS_BINOP_CONDITIONAL_BRANCH (second->opcode) && (first->dreg == second->sregs [0] || first->dreg == second->sregs [1])) {
+		*cond_ins = second;
+		return TRUE;
+	}
+	// pattern `compare; unop conditional branch`
+	if (first && second 
+		&& MINT_IS_COMPARE(first->opcode)
+		&& MINT_IS_UNOP_CONDITIONAL_BRANCH(second->opcode) && first->dreg == second->sregs [0]) {
+		*cond_ins = second;
+		return TRUE;
+	}
+	// pattern `ldc; compare; conditional branch`
+	if (first && second && third 
+		&& MINT_IS_LDC(first->opcode)
+		&& MINT_IS_COMPARE(second->opcode) && (first->dreg == second->sregs [0] || first->dreg == second->sregs [1])
+		&& MINT_IS_UNOP_CONDITIONAL_BRANCH(third->opcode) && second->dreg == third->sregs [0]) {
+		*cond_ins = third;
+		return TRUE;
+	}
+
+	return FALSE;
+}
+
 static void
 interp_reorder_bblocks (TransformData *td)
 {
 	InterpBasicBlock *bb;
+	InterpInst *cond_ins;
 
 	for (bb = td->entry_bb; bb != NULL; bb = bb->next_bb) {
-		InterpInst *first = interp_first_ins (bb);
-		if (!first)
-			continue;
-		if (MINT_IS_CONDITIONAL_BRANCH (first->opcode)) {
-			// This means this bblock has a single instruction, the conditional branch
+		gboolean can_inline_into_callers = interp_inline_into_callers(bb, &cond_ins);
+		if (can_inline_into_callers && MINT_IS_CONDITIONAL_BRANCH(cond_ins->opcode)) {
+			// This means this bblock match a pattern for inlining into callers, with a conditional branch
 			int i = 0;
-			int lookup_var2 = (mono_interp_op_dregs [first->opcode] > 1) ? first->sregs [1] : -1;
 			while (i < bb->in_count) {
 				InterpBasicBlock *in_bb = bb->in_bb [i];
 				InterpInst *last_ins = interp_last_ins (in_bb);
-				if (last_ins && last_ins->opcode == MINT_BR && interp_prev_ins_defines_var (last_ins, first->sregs [0], lookup_var2)) {
+				if (last_ins && last_ins->opcode == MINT_BR && interp_prev_block_defines_var (last_ins, cond_ins)) {
 					// This bblock is reached unconditionally from one of its parents
 					// Move the conditional branch inside the parent to facilitate propagation
 					// of condition value.
-					InterpBasicBlock *cond_true_bb = first->info.target_bb;
+					InterpBasicBlock *cond_true_bb = cond_ins->info.target_bb;
 					InterpBasicBlock *next_bb = bb->next_bb;
 
 					// parent bb will do the conditional branch
 					interp_unlink_bblocks (in_bb, bb);
-					last_ins->opcode = first->opcode;
-					last_ins->sregs [0] = first->sregs [0];
-					last_ins->sregs [1] = first->sregs [1];
+
+					last_ins->opcode = cond_ins->opcode;
+					last_ins->sregs [0] = cond_ins->sregs [0];
+					last_ins->sregs [1] = cond_ins->sregs [1];
 					last_ins->info.target_bb = cond_true_bb;
+					InterpInst *src_ins = interp_prev_ins(cond_ins);
+					InterpInst *dest_ins = interp_prev_ins(last_ins);
+
+					// Copy a pattern into caller bb
+					while (src_ins) {
```

Comment:
This copying here seems a bit hard to follow and unnatural. We could instead just copy all instructions one by one, from `interp_first_ins (bb)` to the end of the `in_bb` (after we got rid of the ending `MINT_BR`).

Reply:
Updated. When copying instructions,`MINT_LDC_I4` has `ins->data [1]` variable and there is a conditional statement for such cases. Other option is to iterate through `ins->data` but it could make the runtime slower.